File: Code/Projects/PyTextFinder/DirNav/dir_nav.py
import os
from collections.abc import Callable


class DirNav:
    _DEFAULT_SKIPS: frozenset[str] = frozenset({
        "bin", "obj",
        "target",
        "build", "out",
        "__pycache__", ".venv", "venv", "dist",
        ".git", ".vs", ".idea",
        "archive",
    })

    def __init__(self, recurse: bool = True) -> None:
        self._recurse = recurse
        self._skips: set[str] = set(self._DEFAULT_SKIPS)
        self._patterns: set[str] = set()
        self._file_count = 0
        self._dir_count = 0
        self.dir_handler: Callable[[str], None] | None = None
        self.file_handler: Callable[[str], None] | None = None

    @property
    def file_count(self) -> int:
        return self._file_count

    @property
    def dir_count(self) -> int:
        return self._dir_count

    def add_pattern(self, ext: str) -> None:
        self._patterns.add(ext.lstrip(".").lower())

    def add_skip(self, name: str) -> None:
        self._skips.add(name)

    def set_recurse(self, r: bool) -> None:
        self._recurse = r

    def visit(self, root: str) -> bool:
        if not os.path.isdir(root):
            return False
        self._file_count = 0
        self._dir_count = 0
        self._visit_impl(root)
        return True

    def _visit_impl(self, dir_path: str) -> None:
        self._dir_count += 1
        normalised = dir_path.replace("\\", "/")
        if handler := self.dir_handler:
            handler(normalised)

        try:
            with os.scandir(dir_path) as it:
                entries = list(it)
        except OSError:
            return

        for entry in entries:
            if entry.is_file(follow_symlinks=False):
                if self._extension_matches(entry.name):
                    self._file_count += 1
                    if handler := self.file_handler:
                        handler(entry.name)
            elif entry.is_dir(follow_symlinks=False):
                if entry.name not in self._skips and self._recurse:
                    self._visit_impl(entry.path)

    def _extension_matches(self, filename: str) -> bool:
        if not self._patterns:
            return True
        _, dot_ext = os.path.splitext(filename)
        ext = dot_ext.lstrip(".").lower()
        return ext in self._patterns

File: Tools/CodeMetrics/TextFinder/CsTextFinder/DirNav/DirNav.cs
namespace DirNav;

public class DirNav
{
    public bool Recurse { get; set; }
    private readonly HashSet<string> _skipList;
    private readonly HashSet<string> _patterns =
        new(StringComparer.OrdinalIgnoreCase);
    private int _fileCount;
    private int _dirCount;

    public Action<string>? DirHandler  { get; set; }
    public Action<string>? FileHandler { get; set; }
    public int FileCount => _fileCount;
    public int DirCount  => _dirCount;

    public DirNav(bool recurse = true)
    {
        Recurse   = recurse;
        _skipList = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            // C#/.NET
            "bin", "obj",
            // Rust
            "target",
            // C++
            "build", "out",
            // Python
            "__pycache__", ".venv", "venv", "dist",
            // common VCS / IDE metadata
            ".git", ".vs", ".idea",
            // archives
            "archive",
        };
    }

    public void AddPattern(string ext) => _patterns.Add(ext.TrimStart('.'));
    public void AddSkip(string name)   => _skipList.Add(name);

    public bool Visit(string root)
    {
        if (!Directory.Exists(root)) return false;
        _fileCount = 0;
        _dirCount  = 0;
        VisitImpl(root);
        return true;
    }

    private void VisitImpl(string dir)
    {
        _dirCount++;
        DirHandler?.Invoke(Normalize(dir));

        try
        {
            foreach (string file in Directory.EnumerateFiles(dir))
            {
                if (ExtensionMatches(file))
                {
                    _fileCount++;
                    FileHandler?.Invoke(Path.GetFileName(file));
                }
            }

            if (Recurse)
            {
                foreach (string sub in Directory.EnumerateDirectories(dir))
                {
                    string name = Path.GetFileName(sub);
                    if (!_skipList.Contains(name))
                        VisitImpl(sub);
                }
            }
        }
        catch (IOException) { }
        catch (UnauthorizedAccessException) { }
    }

    private bool ExtensionMatches(string file)
    {
        if (_patterns.Count == 0) return true;
        string ext = Path.GetExtension(file).TrimStart('.');
        return _patterns.Contains(ext);
    }

    private static string Normalize(string path) => path.Replace('\\', '/');
}
